### ISSUE TITLE
failureRedirect is never used when using with oauth2

### ISSUE THREAD

**Issue by boybundit** [feathersjs/feathers-authentication#387]
### Steps to reproduce

When using with oauth2 plugin with e.g. FacebookStrategy, failureRedirect option is never used by `auth.express.authenticate()` when it receives callback with error e.g. `/auth/facebook/callback?error=access_denied&error_code=200&error_description=Permissions+error&error_reason=user_denied#_=_` and it will throw NotAuthenticated error instead.

### Expected behavior
When callback URL with error is triggered, and failureRedirect is provided, it should redirect accordingly.

### Actual behavior
NotAuthenticated error is thrown.

### System configuration

**Module versions** (especially the part that's not working):
```
+-- feathers@2.0.3
+-- feathers-authentication@1.0.2
+-- feathers-authentication-jwt@0.3.1
+-- feathers-authentication-oauth2@0.2.3
+-- passport-facebook@2.1.1
```

**NodeJS version**: 6.9.1

**Operating System**: Windows 10

**Browser Version**: Chrome 55.0.2883.87 m

**Module Loader**: Node.js

**Comment by boybundit** [comment]
To be specific, in `express/authenticate.js`, why do we need to check `!options.__oauth` here?
```javascript
 if (result.fail) {
        if (options.failureRedirect && !options.__oauth) {
          debug('Redirecting to ' + options.failureRedirect);
          res.status(302);
          return res.redirect(options.failureRedirect);
        }

        var challenge = result.challenge,
            _result$status = result.status,
            status = _result$status === undefined ? 401 : _result$status;

        var message = challenge && challenge.message ? challenge.message : challenge;

        if (options.failureMessage) {
          message = options.failureMessage;
        }

        res.status(status);
        return Promise.reject(new _feathersErrors2.default[status](message, challenge));
      }
```

**Comment by ekryski** [comment]
Weird. That promise rejection should be caught here: https://github.com/feathersjs/feathers-authentication/blob/master/src/express/authenticate.js#L66

I might need to add a few more tests to test redirect. Will take a look over the next few days.

**Comment by ekryski** [comment]
I've verified manually that this is working. Will try and add some more tests today for it.

**Comment by buske** [comment]
I'm experiencing the same issue. The promise is getting caught as expected, and the error is indeed bubbling so that the necessary middleware is executing. However, the feathers-authentication failureRedirect middleware requires `res.hook.data.__redirect` to already be set to the redirect URL (https://github.com/feathersjs/feathers-authentication/blob/master/src/express/failure-redirect.js#L13). 

But `__redirect` is never set in feathers-authentication:authenticate, and is only set by feathers-authentication-oauth2 if the `app.service('authentication').create` call fails (https://github.com/feathersjs/feathers-authentication-oauth2/blob/master/src/express/handler.js#L34-L37). It isn't set when the app.authenticate call fails (https://github.com/feathersjs/feathers-authentication-oauth2/blob/master/src/index.js#L69). I think there needs to be another middleware added to the callback handler to set `__redirect` accordingly so the failureRedirect actually redirects.

**Comment by ekryski** [comment]
Ahhhhh. Good catch @buske!

**Comment by ekryski** [comment]
OAuth2 has been fixed and release as v0.2.4.

Just need to update OAuth1 to close this out.

**Comment by ekryski** [comment]
I started to fix OAuth1 but @buske came through and did it 2 days ago. 👏 .

OAuth1 has been fixed and released as v0.2.4.
